[VPlan] Be more careful with CSE in replicate regions.

fhahn opened the pull request:
Recipes in replicate regions implicitly depend on the region's predicate. Limit CSE to recipes in the same block, when either recipe is in a replicate region.

This allows handling VPPredInstPHIRecipe during CSE. If we perform CSE on recipes inside a replicate region, we may end up with 2 VPPredInstPHIRecipes sharing the same operand. This is incompatible with current VPPredInstPHIRecipe codegen, which re-sets the current value of its operand in VPTransformState. This can cause crashes in the added test cases.

Note that this patch only modifies ::isEqual to check for replicating regions and not getHash, as CSE across replicating regions should be uncommon.

Fixes https://github.com/llvm/llvm-project/issues/157314. 
Fixes https://github.com/llvm/llvm-project/issues/161974.

fhahn committed:
[VPlan] Be more careful with CSE in replicate regions.

Recipes in replicate regions implicitly depend on the region's
predicate. Limit CSE to recipes in the same block, when either recipe is
in a replicate region.

This allows handling VPPredInstPHIRecipe during CSE. If we perform CSE
on recipes inside a replicate region, we may end up with 2
VPPredInstPHIRecipes sharing the same operand. This is incompatible with
current VPPredInstPHIRecipe codegen, which re-sets the current value of
its operand in VPTransformState. This can cause crashes in the added
test cases.

Note that this patch only modifies ::isEqual to check for replicating
regions and not getHash, as CSE across replicating regions should be uncommon.

Fixes https://github.com/llvm/llvm-project/issues/157314.
Fixes https://github.com/llvm/llvm-project/issues/161974.

artagnon reviewed (commented):
> Note that this patch only modifies ::isEqual to check for replicating regions and not getHash, as CSE across replicating regions should be uncommon.

Is there some issue with hashing the parent region in the case of replicates? Would be safer to do that?

artagnon reviewed (commented):
> If we perform CSE on recipes inside a replicate region, we may end up with 2 VPPredInstPHIRecipes sharing the same operand.

Hm, this reminds me of our fix to the const-folder. I guess the cse and const-folder are more similar than I thought (also because I have a proposed patch to share getOpcode).

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
I think it would be cleaner to handle PredPHI here, as I think getOpcodeOrIntrinsicID can be shared?

fhahn replied:
I don't think that's going to work, as we need to return a proper opcode from `getOpcodeOrIntrinsicID`. Currently we return `std::nullopt` for VPVectorPointerRecipe, which works for now because there is only a single recipe we return that for, but needs to be fixed to support more recipes that don't map directly to llvm::Instruction opcodes: https://github.com/llvm/llvm-project/pull/162267

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```suggestion
        Result = hash_combine(Result, RFlags->getPredicate());
    const VPRegionBlock *RepR = Def->getParent()->getParent();
    if (RepR && RepR->isReplicator())
      Result = hash_combine(Result, Def->getParent());
```

fhahn committed:
[VPlan] Assign custom opcodes to recipes not mapping to IR opcodes.

We can perform CSE on recipes that do not directly map to Instruction
opcodes. One example is VPVectorPointerRecipe. Currently this is handled
by supporting them in ::canHandle, but currently that means that we
return std::nullopt from getOpcodeOrIntrinsicID() for it. This currently
only works, because the only case we return std::nullopt and perform CSE
is VPVectorPointerRecipe. But that does not work if we support more such
recipes, like VPPredInstPHIRecipe
(https://github.com/llvm/llvm-project/pull/162110).

To fix this, return a custom opcode from getOpcodeOrIntrinsicID for
recipes like VPVectorPointerRecipe, using the VPDefID after all regular
instruction opcodes.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-cse-replicator

fhahn committed:
!fixup use DefID

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-cse-replicator

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-cse-replicator

artagnon reviewed (approved):
LGTM, thanks! Thanks for the explanation: should be good.

fhahn committed:
Merge branch 'main' into vplan-cse-replicator
